src/main/java/org/gradle/profiler/asyncprofiler/AsyncProfilerDownload.java

```diff
@@ -69,28 +79,50 @@ static File defaultHome() {
         return homeDir;
     }
 
+    private static void unzipTo(File source, File destDir) throws IOException {
+        try (ZipArchiveInputStream zipStream = new ZipArchiveInputStream(new FileInputStream(source))) {
+            extract(zipStream, destDir.toPath());
+        }
+    }
+
     private static void untarTo(File source, File destDir) throws IOException {
         try (TarArchiveInputStream tarStream = new TarArchiveInputStream(new GzipCompressorInputStream(new FileInputStream(source)))) {
-            while (tarStream.getNextEntry() != null) {
-                if (tarStream.getCurrentEntry().isDirectory()) {
-                    continue;
-                }
-                String name = tarStream.getCurrentEntry().getName();
-                File file = new File(destDir, name);
-                Files.createDirectories(file.getParentFile().toPath());
-                Files.copy(tarStream, file.toPath(), REPLACE_EXISTING);
-                boolean executable = (tarStream.getCurrentEntry().getMode() & 0x40) != 0;
-                if (executable) {
-                    Set<PosixFilePermission> permissions = Files.getPosixFilePermissions(file.toPath());
-                    ImmutableSet.Builder<PosixFilePermission> withExecute = ImmutableSet.builder();
-                    withExecute.addAll(permissions);
-                    withExecute.add(PosixFilePermission.OWNER_EXECUTE);
-                    Files.setPosixFilePermissions(file.toPath(), withExecute.build());
-                }
+            extract(tarStream, destDir.toPath());
+        }
+    }
+
+    private static void extract(ArchiveInputStream archiveStream, Path destDir) throws IOException {
+        ArchiveEntry entry;
+        while ((entry = archiveStream.getNextEntry()) != null) {
+            if (entry.isDirectory()) {
+                continue;
+            }
+            String name = entry.getName();
+            Path file = destDir.resolve(name);
+            if (!file.startsWith(destDir)) {
```

Comment:
I think this might fix it:
```suggestion
            if (!file.normalize().startsWith(destDir)) {
```

I tried it in https://github.com/gradle/gradle-profiler/pull/408 and no issue is reported.